lib/battle/Unit.cpp

```diff
@@ -0,0 +1,220 @@
+/*
+ * Unit.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#include "StdInc.h"
+
+#include "Unit.h"
+
+#include "../VCMI_Lib.h"
+#include "../CGeneralTextHandler.h"
+#include "../NetPacksBase.h"
+
+#include "../serializer/JsonDeserializer.h"
+#include "../serializer/JsonSerializer.h"
+
+namespace battle
+{
+
+///Unit
+bool Unit::isDead() const
+{
+	return !alive() && !isGhost();
+}
+
+bool Unit::isTurret() const
+{
+	return creatureIndex() == CreatureID::ARROW_TOWERS;
+}
+
+bool Unit::isValidTarget(bool allowDead) const
+{
+	return (alive() || (allowDead && isDead())) && getPosition().isValid() && !isTurret();
+}
+
+std::string Unit::getDescription() const
+{
+	boost::format fmt("Unit %d of side %d");
+	fmt % unitId() % unitSide();
+	return fmt.str();
+}
+
+
+std::vector<BattleHex> Unit::getSurroundingHexes(BattleHex assumedPosition) const
+{
+	BattleHex hex = (assumedPosition != BattleHex::INVALID) ? assumedPosition : getPosition(); //use hypothetical position
+	std::vector<BattleHex> hexes;
+	if(doubleWide())
+	{
+		const int WN = GameConstants::BFIELD_WIDTH;
+		if(unitSide() == BattleSide::ATTACKER)
+		{
+			//position is equal to front hex
+			BattleHex::checkAndPush(hex - ((hex / WN) % 2 ? WN + 2 : WN + 1), hexes);
+			BattleHex::checkAndPush(hex - ((hex / WN) % 2 ? WN + 1 : WN), hexes);
+			BattleHex::checkAndPush(hex - ((hex / WN) % 2 ? WN : WN - 1), hexes);
+			BattleHex::checkAndPush(hex - 2, hexes);
+			BattleHex::checkAndPush(hex + 1, hexes);
+			BattleHex::checkAndPush(hex + ((hex / WN) % 2 ? WN - 2 : WN - 1), hexes);
+			BattleHex::checkAndPush(hex + ((hex / WN) % 2 ? WN - 1 : WN), hexes);
+			BattleHex::checkAndPush(hex + ((hex / WN) % 2 ? WN : WN + 1), hexes);
+		}
+		else
+		{
+			BattleHex::checkAndPush(hex - ((hex / WN) % 2 ? WN + 1 : WN), hexes);
+			BattleHex::checkAndPush(hex - ((hex / WN) % 2 ? WN : WN - 1), hexes);
+			BattleHex::checkAndPush(hex - ((hex / WN) % 2 ? WN - 1 : WN - 2), hexes);
+			BattleHex::checkAndPush(hex + 2, hexes);
+			BattleHex::checkAndPush(hex - 1, hexes);
+			BattleHex::checkAndPush(hex + ((hex / WN) % 2 ? WN - 1 : WN), hexes);
+			BattleHex::checkAndPush(hex + ((hex / WN) % 2 ? WN : WN + 1), hexes);
+			BattleHex::checkAndPush(hex + ((hex / WN) % 2 ? WN + 1 : WN + 2), hexes);
```

Comment:
It will be great use here some existing things like get 2x times BattleHex::getNeighboringTiles() and then merge them without positions hexes. Eventually  just use cloneInDirection in checkAndPush instead calculations.

Reply:
Rewritten with `cloneInDirection `